### ISSUE TITLE
multiple call to mbedtls_ssl_handshake on server create a memory leak

### ISSUE THREAD

**Issue by giso-c** [ARMmbed/mbedtls#4071]
### Description
- Type: Bug
- Priority: Blocker

---------------------------------------------------------------
## Bug

when calling multiple time mbedtls_ssl_handshake, it will create a memory leak

**OS**  
Freertos

**mbed TLS build:**    
TLS 2.25
Following define are used as we are using multiple threads
#define MBEDTLS_THREADING_C
#define MBEDTLS_THREADING_ALT

**Expected behavior**   
We have a TLS server, upon client connection we call mbedtls_ssl_handshake. Once the connection is terminated FW waits for a new client connection and call again mbedtls_ssl_handshake.

mbedtls_ssl_handshake calls multiple times
mbedtls_ctr_drbg_init
mbedtls_ctr_drbg_free
mbedtls_ctr_drbg_init
mbedtls_ctr_drbg_free

Issue is that at the end of mbedtls_ssl_handshake we have 1 mutex created and it will never be deleted so next call to mbedtls_ssl_handshake will add a new mutex.

There is only 1 way to delete the mbedtls_ctr_drbg_context mutex is by calling mbedtls_ctr_drbg_free, but mbedtls_ctr_drbg_free will create a new mutex.

To reproduce,  call multiple times mbedtls_ssl_handshake with MBEDTLS_THREADING_C and MBEDTLS_THREADING_ALT defined

**Comment by gilles-peskine-arm** [comment]
I guess this is the same problem as https://github.com/ARMmbed/mbedtls/issues/4045? Can you try to see if Mbed TLS 2.24.0 works?

**Comment by giso-c** [comment]
If I remove mutex creation from mbedtls_ctr_drbg_free  like this

void mbedtls_ctr_drbg_free( mbedtls_ctr_drbg_context *ctx )
{
    if( ctx == NULL )
        return;

#if defined(MBEDTLS_THREADING_C)
    mbedtls_mutex_free( &ctx->mutex );
#endif
    mbedtls_aes_free( &ctx->aes_ctx );
    mbedtls_platform_zeroize( ctx, sizeof( mbedtls_ctr_drbg_context ) );
    ctx->reseed_interval = MBEDTLS_CTR_DRBG_RESEED_INTERVAL;
    ctx->reseed_counter = -1;
#if defined(MBEDTLS_THREADING_C)
    // mbedtls_mutex_init( &ctx->mutex );
#endif
}

It works, this is similar to version 2.24 so I believe2.24 should also work but not sure if this will create some other issue.

Kind Regards ǀ Mit freundlichen Grüßen ǀ Med<http://browse.dict.cc/norwegisch-deutsch/Med.html> vennlig<http://browse.dict.cc/norwegisch-deutsch/vennlig.html> hilsen<http://browse.dict.cc/norwegisch-deutsch/hilsen.html>

Adalgiso Castrignano ǀ Senior Embedded Software Engineer

IDENTEC SOLUTIONS AG
Millennium Park 2 ǀ 6890 Lustenau ǀ A U S T R I A

T +43 5577 87387-64

acastrignano@identecsolutions.com<mailto:acastrignano@identecsolutions.com>
https://www.identecsolutions.com/

Data Privacy Statement / Datenschutzerklärung<https://www.identecsolutions.com/data-privacy/>

[cid:image001.jpg@01D6F4AC.D4DAA720][cid:image005.jpg@01D6F4AC.D4E92610]

Visibility Delivered.
________________________________________
This email message may contain confidential information. If you are not the intended recipient, any use, dissemination, distribution or copying of this email message is strictly prohibited. If you have received this message in error, please immediately notify the sender and delete this email message from your computer.

Firmenbuch Feldkirch    FN:181501p


IDENTEC SOLUTIONS AG has always taken your privacy seriously and with the new laws in place regarding your data privacy rights (EU General Data Protection Regulation), we thought it is a good time to remind you of our commitment. You can access our data privacy statement<https://www.identecsolutions.com/data-privacy/> in full on our website.

**Comment by giso-c** [comment]
I  believe it would be hard for you to tell when this issue will be evaluated and fixed.
Do you think it would be better to use 2.24 or to use the temporary fix (not creating mutex when calling mbedtls_ctr_drbg_free)?

**Comment by chris-jones-arm** [comment]
Hi @giso-c, thanks for the bug report! After a quick look this issue does look like a duplicate of #4045 however I will keep this open for now until this can be proven.

I would advise against using a temporary fix unless you absolutely need a new feature or bug fix that was released in 2.25 as removing that mutex could have unforseen consequences which are not immediately obvious. If you can stick with 2.24 then that should be preferred until this issue is fixed.

**Comment by gilles-peskine-arm** [comment]
If you decide to stick with 2.24, please review the [security issues in the ChangeLog file](https://github.com/ARMmbed/mbedtls/blob/development/ChangeLog). Fortunately in this release all the security fixes are relatively minor in the sense that they're either only about a second line of defense or about a rarely-used feature, but do review what might apply to you.

**Comment by gilles-peskine-arm** [comment]
There's a proposed fix in https://github.com/ARMmbed/mbedtls/pull/4090. Please note that it has not been reviewed yet, and we may yet revise how we resolve this issue (it's not fully clear to me how to fix this issue without breaking “unusual” usage of the API). If you want to try it out, you can apply the patches from https://github.com/gilles-peskine-arm/mbedtls/tree/fix-drbg-mutex-usage-2.25-untested.

**Comment by giso-c** [comment]
I got the latest code, but did not fix the memory leak, same code running on 2.14 works fine but new code does not work.
I checked the heap before and after calling

**Comment by gilles-peskine-arm** [comment]
Thanks for checking. I've started to test the 2.7 patches on `development` and I can reproduce the leak (`tests/ssl-opt.sh -f Default -m` passes on my 2.7 fix branch, but not yet on my development fix branch).

**Comment by gilles-peskine-arm** [comment]
Actually, I can't reproduce the problem. The leak I saw was some missing cleanup in the test program. Can you share your exact compile-time configuration and a precise way to reproduce the problem? Either with a small, self-contained program that can build on Linux (even if you only see the leak on FreeRTOS — mutexes on Linux don't consume extra heap, which is why we didn't see notice the bug immediately), or with `programs/ssl/ssl_client2` or `programs/ssl/ssl_server2` (which are the programs I'm using for testing).

**Comment by gilles-peskine-arm** [comment]
Oh, and please test with https://github.com/gilles-peskine-arm/mbedtls/tree/test-mutex-usage-count-development rather than the branch I posted earlier. I don't think I made any relevant change (the differences should only be in test code), but you never know.

**Comment by gilles-peskine-arm** [comment]
I'm closing this issue as a duplicate of https://github.com/ARMmbed/mbedtls/issues/4017. It's either exactly the same problem (maybe in a different branch) or a slight variation (e.g. HMAC_DRBG vs CTR_DRBG) which will be fixed together. Please reopen this issue once the proposed fix is merged if it doesn't fix your problem.

We can continue exchanging comments about the proposed fix on this thread.

**Comment by gilles-peskine-arm** [comment]
I'm closing this issue as a duplicate of https://github.com/ARMmbed/mbedtls/issues/4017. It's either exactly the same problem (maybe in a different branch) or a slight variation (e.g. HMAC_DRBG vs CTR_DRBG) which will be fixed together. Please reopen this issue once the proposed fix is merged if it doesn't fix your problem.

We can continue exchanging comments about the proposed fix on this thread.
